### PR TITLE
schema.org: Remove default type "Product"

### PR DESCRIPTION
**Pull request by mtrojan-ub**
`Product`
> Any offered product or service. For example: a pair of shoes; a concert ticket; the rental of a car; a haircut; or an episode of a TV show streamed online.

https://schema.org/Product

VuFind is often used as a Bibliography with the main goal to provide Metadata and maybe also provide links to access full texts via external websites. Of course, there can be cases where you provide access by yourself, but that does not mean that every Record should be marked as `Product` by default. Google will even treat all pages as erroneous that are marked as `Product` but have neither `offers` nor `review` nor `aggregateRating` information available.

https://developers.google.com/search/docs/appearance/structured-data/product-snippet?hl=en#structured-data-type-definitions

I don't think this a good final solution, but i definitely think that `Product` should not be hardcoded in the default templates.

Note: It would be better to let the RecordDriver (getSchemaOrgFormatsArray) handle the decision whether this is a Product or not, but what would be a proper default mechanism?

### CODE REVIEW

**Review by demiankatz — CHANGES_REQUESTED** [review]
@mtrojan-ub, VuFind's schema.org implementation uses the [Offer](https://schema.org/Offer) schema to represent availability. My understanding (though it may be based on incomplete or outdated information) is that the Offer can only be applied to a Product, which is why the Product class is mandatory. Simply removing it may completely break the representation of Offers. Maybe it's worth carefully studying the linked documentation to see if I'm mistaken or if we should add additional properties or make other adjustments to make the representation more robust or reliable.

Since there are known problems with the current implementation, as you note, #3634 was created to offer the ability to completely disable schema.org for situations where it is not working well (this was included in release 9.0). It may be best for you to use that setting until we can figure out a way to make the overall representation more robust.

I'll leave this open for further discussion -- I'd love to find a solution but haven't had time to do a deep dive into the linked data. However, I don't think we can merge it as-is.

**Comment by mtrojan-ub** [comment]
Thanks for your statement, @demiankatz!

We already made changes to our custom templates, there has just been some regression during the upgrade to 9.1, but we fixed that for now. The reason i opened this PR is just that i thought others might have a similar problem and it might be worth changing something in case others run into the same issue.

Just one more thought: If i understand correctly, the **Offer** schema is bound to the ILS functionality. Maybe it would be a simple solution to remove the Product tag if the NoILS driver is used?

**Comment by demiankatz** [comment]
> Just one more thought: If i understand correctly, the **Offer** schema is bound to the ILS functionality. Maybe it would be a simple solution to remove the Product tag if the NoILS driver is used?

That's a possibility, though it would have to be based on more than just the use of the NoILS driver, since in some modes, NoILS will simulate a real ILS and provide item availability data based on MARC fields. I think the appropriate check would be `if ($this->ils()->getOfflineMode() !== 'ils-none'):` to add the class. Maybe for clarity we would want to add some new method to the schemaOrg helper to encapsulate this logic (like `getDefaultRecordTypes` or something).

**Comment by mtrojan-ub** [comment]
Good point, @demiankatz! I just changed the logic & moved it into the SchemaOrg helper. I even think it might be a good idea to put the logic to combine the default + record specific types into the helper as well. This way we might add more types / resolve conflicts between types inside the helper if this should ever be necessary at some point, without changing the templates. What do you think?

File: module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php [hunk 1]
```diff
@@ -106,4 +111,53 @@ public function getMeta(string $property, string $content, array $attributes = [
...
     }
+
+    /**
+     * Get default record types as array.
+     *
+     * @return array
+     */
+    public function getDefaultRecordTypesArray(): array
+    {
+        if ($this->ils->__invoke()->getOfflineMode() !== 'ils-none') {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php]: For more concise syntax, you can replace `$this->ils->__invoke()->...` with `($this->ils)()->...`. Just note the need for parentheses around `$this->ils` to allow it to be evaluated as a function.

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordDataFormatterTest.php [hunk 2]
```diff
@@ -110,7 +110,8 @@ protected function getViewHelpers($container): array
             'record' => $record,
             'recordLinker' => new \VuFind\View\Helper\Root\RecordLinker($this->getMockRecordRouter()),
             'schemaOrg' => new \VuFind\View\Helper\Root\SchemaOrg(
-                new \Laminas\View\Helper\HtmlAttributes()
+                new \Laminas\View\Helper\HtmlAttributes(),
+                $this->getMockBuilder(\VuFind\View\Helper\Root\Ils::class)->disableOriginalConstructor()->getMock()
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordDataFormatterTest.php]: You can use `$this->createMock(\VuFind\View\Helper\Root\Ils::class)` to be more concise while achieving the same result as using the mock builder.

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/SchemaOrgTest.php [hunk 3]
```diff
@@ -52,7 +53,8 @@ class SchemaOrgTest extends \PHPUnit\Framework\TestCase
      */
     public function getHelper(bool $enabled): SchemaOrg
     {
-        return new SchemaOrg(new HtmlAttributes(), $enabled);
+        return new SchemaOrg(new HtmlAttributes(), $this->getMockBuilder(Ils::class)
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/SchemaOrgTest.php]: Another spot where using `createMock` would be more concise.

File: module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php [hunk 4]
```diff
@@ -106,4 +111,53 @@ public function getMeta(string $property, string $content, array $attributes = [
...
+     *
+     * @param DefaultRecord $driver Record Driver
+     *
+     * @return array
+     */
+    public function getRecordTypesArray(DefaultRecord $driver): array
+    {
+        $defaultTypes = $this->getDefaultRecordTypesArray();
+        $recordTypes = $driver->getSchemaOrgFormatsArray();
+        return array_merge($defaultTypes, $recordTypes);
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php]: Maybe it's worth adding an `array_unique` around this to avoid duplication (just to be on the safe side).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @mtrojan-ub, this looks good to me, and I agree that doing the combining inside the helper helps to keep the templates simple and straightforward.

I just have a few trivial suggestions below. It might also be worth adding a couple of new tests to the helper unit test to cover the new methods. I'm willing to help with that if you'd like me to.

Once that's sorted out, the only remaining question is whether the changed helper constructor signature is reason enough to defer this to release 11, or if it's okay to merge this into 10.1 with a changelog note. I think it's probably okay for 10.1 given that this is a relatively new helper and there hasn't historically been much reason for anyone to customize/override it.

**Comment by mtrojan-ub** [comment]
Thanks, @demiankatz! Your feedback has been implemented. Of course i'm always happy if you're willing to help regarding the unit tests! I also think that 10.1 should be fine (if possible) since the amount of changed files is very small.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Sorry for not thinking of this sooner, but when I started working on unit tests for this, I realized that there might be a better way to approach it.

My initial thought was that we might want to deprecate the record driver's getSchemaOrgFormats() method, since getSchemaOrgFormatsArray() provides all the data, and the formatting work is probably better addressed by the schemaOrg helper. I still think that may be the case.

However, when I started looking through the code to see where getSchemaOrgFormats() was used, I discovered that there's schema.org logic in several templates beyond the ones being modified in this PR -- specifically, the EDS and EPF core templates. We probably ought to adjust those templates to use the helper consistently with the default core template.

But that made me realize that checking the ILS status in the helper isn't really correct, because we don't know whether the record being examined belongs to a backend that corresponds to the ILS, or some external service. The record driver itself knows this, so I think more logic needs to be merged into the record driver, instead of doing the work in the schemaOrg helper.

In other words, I still think it's a good idea to create the getRecordTypes method in the schemaOrg helper... but I think we can remove the merging and ILS-specific logic from the helper (which also gets rid of the backward-incompatible constructor change) and instead add something like this to DefaultRecord::getSchemaOrgFormatsArray():

```
// Check for functionality from IlsAwareTrait. If this record comes from a real ILS, we need
// to add a type of "Product" to support the system's use of https://schema.org/Offer to
// represent availability.
if ($this->tryMethod('hasILS') && isset($this->ils) && $this->ils->getOfflineMode() !== 'ils-none') {
    $types[] = 'Product';
}
```

What do you think?

**Comment by mtrojan-ub** [comment]
This might indeed be the better+easier solution, especially if it also covers the EDS/EPF cases!

File: module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php [hunk 5]
```diff
@@ -106,4 +109,28 @@ public function getMeta(string $property, string $content, array $attributes = [
...
     }
+
+    /**
+     * Get all record types for the given record.
+     *
+     * @param DefaultRecord $driver Record Driver
+     *
+     * @return array
+     */
+    public function getRecordTypesArray(DefaultRecord $driver): array
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php]: It might be more open-ended to use \VuFind\RecordDriver\AbstractBase as the type instead of DefaultRecord, and then change the code to: `return $driver->tryMethod('getSchemaOrgFormatsArray') ?? [];` to be sure that nothing strange happens if an exotic record driver gets passed to the helper. (I think that's unlikely to happen, but it doesn't hurt to prepare for the possibility).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @mtrojan-ub, I think this looks very good -- just one remaining suggestion, and once that's adjusted, I'll add some test coverage.

File: module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php (hunk 5 above)
Comment (demiankatz) [reply module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php]: I had a few minutes this afternoon, so I went ahead and updated the unit test for the SchemaOrg helper. I realized that doing this required my above suggested changes since the test harness for record drivers is not a DefaultRecord subclass, so I went ahead and applied those changes. Please take a look and let me know what you think!

Comment (mtrojan-ub) [reply module/VuFind/src/VuFind/View/Helper/Root/SchemaOrg.php]: Thanks, looks good to me!

**Review by demiankatz — APPROVED** [review]
Thanks again, @mtrojan-ub! I've made a few more minor tweaks (adding more test coverage, deprecating the now-obsolete record driver method so it can be removed in 11.0). I'll merge as soon as continuous integration passes.
